### ISSUE TITLE
BCER eval displayed during lstmtraining and that from lstmeval are different

### ISSUE THREAD

**Issue by Shreeshrii** [tesseract-ocr/tesseract#3763]
While trying to plot the error rates for training, I have come across an anomaly.

I use the LOG file generated from messages output during lstmtraining run, which also out `BCER eval` on completion of evaluation using the eval list. This message displays the error rates as well as the learning iteration.

I separately run lstmeval on various checkpoint traineddata files to get the error rates.

 I have found that BCER eval displayed during lstmtraining and that from lstmeval are different for the same `learning teration'.

**Comment by Shreeshrii** [comment]
Below are plots from a recent training, trying to add superscripts to the English traineddata.

The first is the chart generated from the lstmtraining log file, plotting the BCER every 100 iterations, every checkpoint and every eval. Since `BCER eval` only reports the learning iteration value, the main x axis plots learning iterations.

![engSuper-LOG-2](https://user-images.githubusercontent.com/5095331/156589377-3dc1ff90-a4c7-4a99-bca3-fd7ff718fc90.png)

The second chart plots the CER values calculated for fast traineddata files for checkpoints with BCER less than 1% using lstmeval, ISRI OCR evaluation and OCRevaluation in addition to BCER from lstmtraining at every 100 iterations and every checkpoint. It uses training iterations as the main x axis and hence does not include the eval done during training (included in the chart above).

![engSuper-2](https://user-images.githubusercontent.com/5095331/156590414-8fd5953a-db6e-4918-8206-9549bb8a31be.png)

**Comment by Shreeshrii** [comment]
Here is the tsv file with lstmeval BCER for checkpoints with BCER less than 1%.

|Name|CheckpointCER|LearningIteration|TrainingIteration|EvalCER |IterationCER|SubtrainerCER|
|----|-------------|-----------------|-----------------|--------|------------|-------------|
|    |             |26767            |44800            |1.573309|            |             |
|    |             |26795            |44900            |1.616475|            |             |
|    |             |29730            |53300            |1.398542|            |             |
|    |             |33220            |63700            |1.312008|            |             |
|    |             |37424            |77200            |1.227363|            |             |
|    |             |37651            |78000            |1.346600|            |             |
|    |             |37739            |78300            |1.228998|            |             |
|    |             |37823            |78600            |1.219876|            |             |
|    |             |41319            |90900            |1.113027|            |             |
|    |             |44647            |103100           |1.189317|            |             |
|    |             |47598            |114700           |1.177244|            |             |
|    |             |49613            |123000           |1.243091|            |             |
|    |             |50527            |126700           |1.010922|            |             |
|    |             |55098            |146200           |0.920837|            |             |
|    |             |55708            |149100           |1.006433|            |             |
|    |             |55750            |149300           |0.920407|            |             |
|    |             |63131            |184300           |0.919421|            |             |
|    |             |64733            |192400           |0.954834|            |             |
|    |             |69690            |218000           |0.842977|            |             |
|    |             |70556            |222600           |0.976068|            |             |
|    |             |74289            |242800           |0.864550|            |             |
|    |             |76879            |257300           |0.795503|            |             |
|    |             |85113            |304600           |0.720793|            |             |
|    |             |87091            |317000           |0.747304|            |             |
|    |             |87767            |321100           |0.928360|            |             |
|    |             |88511            |325900           |0.956177|            |             |
|    |             |91252            |343500           |0.989352|            |             |
|    |             |99537            |398000           |0.768607|            |             |
|    |             |110002           |472900           |0.914486|            |             |
|    |             |112642           |493200           |0.951797|            |             |
|    |             |126077           |598400           |0.822581|            |             |
|    |             |127313           |608800           |0.762501|            |             |
|    |             |130388           |635200           |0.701150|            |             |
|    |             |130395           |635300           |0.754357|            |             |
|    |             |145368           |766900           |1.017711|            |             |
|    |             |161436           |921200           |0.722449|            |             |
|    |             |161990           |926800           |0.744561|            |             |
|    |             |161997           |926900           |0.746143|            |             |
|    |             |170477           |1012400          |0.676673|            |             |
|    |             |201343           |1356000          |0.821650|            |             |
|    |             |201356           |1356200          |0.811972|            |             |

**Comment by Shreeshrii** [comment]
So, for the checkpoint with minimal BCER during training, the values are:

checkpoint training iterations = 1356200
checkpoint learning iterations = 201356
checkpoint BCER = 0.110

lowest eval learning iterations = 200061
lowest eval BCER = 0.628007	

eval learning iterations = 201356
eval BCER = 0.794562

lstmeval training iterations = 1356200
lstmeval learning iterations = 201356
lstmeval BCER = 0.811972

As shown above, there is a difference in the lstmeval BCER (0.811972) compared to the BCER eval reported for the same number of learning iterations (0.794562). In my opinion, both should give the same result.

So, there seems to be some error in the reporting of learning iterations # for BCER eval. Maybe it would help, if both training iterations and learning iterations are reported for BCER eval done during lstmtraining.

**Comment by amitdo** [comment]
Does both lstmtraining and lstmeval evaluate exactly the same subset of images+ground truth and in the same order in each evaluation cycle?

**Comment by Shreeshrii** [comment]
Both evaluate the same subset of images+ground truth, the ones listed in list.eval. training is done on list.train. lstmeval run on checkpoints uses list.eval and goes through it sequentially. I do not know how the order of files for eval is decided during lstmtraining. 

I also do not know what is the learning iteration number reported by eval during lstmtraining, i.e. whether it is the current learning iteration number when the eval is being reported or whether it is the saved iteration number from when eval was started

**Comment by whisere** [comment]
having the same issue, this might be related to https://github.com/tesseract-ocr/tesstrain/issues/110 ?

**Comment by bertsky** [comment]
I can confirm this is still an issue. In my case, the difference is much worse:

| plot_log | plot_cer |
| --- | --- |
| ![htrbin again plot_log fast](https://github.com/tesseract-ocr/tesseract/assets/38561704/88e02db9-9d7e-41e0-8424-ad14fbd9c4cd) | ![htrbin again plot_cer fast](https://github.com/tesseract-ocr/tesseract/assets/38561704/3a0644ab-5771-436d-bf9c-e5032ff3ac16) |

As you can see, the lstmeval BCER is close to 100%, while the lstmtraining BCER is around 11%.

If I replace `fast` models with `best` models for checkpoint extraction in the rules for `make plot`, then the difference becomes benign:

| plot_log | plot_cer |
| --- | --- |
| ![htrbin again plot_log](https://github.com/tesseract-ocr/tesseract/assets/38561704/18c4fca1-183c-4b5d-900e-2ca464153e13)| ![htrbin again plot_cer](https://github.com/tesseract-ocr/tesseract/assets/38561704/df9c97d6-fce6-4afe-a737-39bb037e7047) |

So for me the observation that `convert_to_int` is the culprit seems true.

Another example pointing in a similar direction (different Tesseract/Tesstrain installation, different data):

| plot_log | plot_cer |
| --- | --- |
| ![herrnhut-kurrent tess bin-1CNN-Lfx400 broken plot_log](https://github.com/tesseract-ocr/tesseract/assets/38561704/d7e19888-8622-4289-a826-fb17f40a30aa) | ![herrnhut-kurrent tess bin-1CNN-Lfx400 broken plot_cer](https://github.com/tesseract-ocr/tesseract/assets/38561704/90961b6b-d652-43c7-8456-91fa32ced4c0) |

And again with best instead of fast extraction:

| plot_log | plot_cer |
| --- | --- |
| ![herrnhut-kurrent tess bin-1CNN-Lfx400 plot_log](https://github.com/tesseract-ocr/tesseract/assets/38561704/57630c82-df49-4a10-bc61-eb8d2542728a) | ![herrnhut-kurrent tess bin-1CNN-Lfx400 plot_cer](https://github.com/tesseract-ocr/tesseract/assets/38561704/b000fe90-b08a-4a3f-b23b-45f371590fd0) |

Note that if I apply the models directly with the tesseract CLI, I can reproduce the behaviour shown in the plots – results are gibberish with the fast models, but ok with the best models.

**Comment by amitdo** [comment]
The net spec for 'best' and 'fast' is not the same.

Any 'fast' model is basically another 'best' model that was converted to int. The 'best' of 'fast' was never release publicly.

**Comment by amitdo** [comment]
https://github.com/tesseract-ocr/tessdoc/blob/441f1ea328421e/Data-Files-in-tessdata_best.md#version-string--40000alpha--network-specification-for-tessdata_best

https://github.com/tesseract-ocr/tessdoc/blob/441f1ea328421e/Data-Files-in-tessdata_fast.md

**Comment by bertsky** [comment]
@amitdo 
> The net spec for 'best' and 'fast' is not the same.

That's not true. The [VGSL spec / net_mode](https://tesseract-ocr.github.io/tessdoc/tess4/VGSLSpecs.html) is the same, only the extraction method differs.

Anyway, that's irrelevant here, since the problem appears independent of where the training started (pretrained models in the tessdata repos or from scratch). The relevant difference is the checkpoint extraction method.

BTW, I am not saying `fast` _always_ behaves like this, it's still somewhat surprising. I guess it depends on the course taken during lstmtraining – perhaps subtrainer feedback or other events. But users should avoid the `fast` method for now to be on the safe side IMO.

@Shreeshrii have you by any chance noticed any particular event in the training log which we can use to track this down?
